File: app/src/main/java/com/dongumen/nickolas/kpiweeks/widgets/adapters/delegateAdapter/BaseDelegateAdapter.java
package com.dongumen.nickolas.kpiweeks.widgets.adapters.delegateAdapter;

import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import org.jetbrains.annotations.NotNull;

import java.util.List;

public abstract class BaseDelegateAdapter
        <VH extends BaseViewHolder, T> implements IDelegateAdapter<VH, T> {

    abstract protected void onBindViewHolder(
            @NonNull View view, @NonNull T item, @NonNull VH viewHolder);

    @LayoutRes
    abstract protected int getLayoutId();

    @NonNull
    abstract protected VH createViewHolder(View parent);

    @Override
    public void onRecycled(VH holder) {
    }

    @NonNull
    @Override
    public final RecyclerView.ViewHolder onCreateViewHolder(
            @NonNull ViewGroup parent, int viewType) {

        final View inflatedView = LayoutInflater
                .from(parent.getContext())
                .inflate(getLayoutId(), parent, false);
        final VH holder = createViewHolder(inflatedView);
        holder.setListener((viewType1, view) -> onBindViewHolder(view, (T) viewType1, holder));
        return holder;
    }

    @Override
    public void onBindViewHolder(@NotNull VH holder, @NotNull List<? extends T> items, int position) {
        holder.bind(items.get(position));
    }
}

File: app/src/main/java/com/dongumen/nickolas/kpiweeks/global/delegateAdapter/BaseDelegateAdapter.kt
package com.dongumen.nickolas.kpiweeks.global.delegateAdapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.LayoutRes
import androidx.recyclerview.widget.RecyclerView

abstract class BaseDelegateAdapter<VH : BaseViewHolder, T : Any> : IDelegateAdapter<VH, T> {

    @get:LayoutRes
    protected abstract val layoutId: Int

    protected abstract fun onBindViewHolder(
            view: View, item: T, viewHolder: VH)

    protected abstract fun createViewHolder(parent: View): VH

    override fun onRecycled(holder: VH) {}

    override fun onCreateViewHolder(
            parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {

        val inflatedView = LayoutInflater
                .from(parent.context)
                .inflate(layoutId, parent, false)
        val holder = createViewHolder(inflatedView)
        holder.setListener(object : BaseViewHolder.ItemInflateListener {
            override fun inflated(viewType: Any, view: View) {
                onBindViewHolder(view, viewType as T, holder)
            }
        })
        return holder
    }

    override fun onBindViewHolder(holder: VH, items: List<T>, position: Int) {
        holder.bind(items[position])
    }
}
